sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/XXE/XmlReaderSettingsValidator.cs

```diff
@@ -77,15 +106,27 @@ public bool IsUnsafe(InvocationExpressionSyntax invocation, ISymbol settings, Ne
                 .SelectMany(location => GetDescendantNodes(location, invocation).OfType<ObjectCreationExpressionSyntax>())
                 .FirstOrDefault(objectCreation => objectCreation.Initializer != null && IsXmlReaderSettings(objectCreation, semanticModel));
 
-        private static IEnumerable<SyntaxNode> GetDescendantNodes(Location location, InvocationExpressionSyntax invocation) =>
-            location.SourceTree?.GetRoot()?.FindNode(location.SourceSpan).DescendantNodes() ??
-            invocation.FirstAncestorOrSelf<MethodDeclarationSyntax>()?.DescendantNodes() ??
-            Enumerable.Empty<SyntaxNode>();
+        private static IEnumerable<SyntaxNode> GetDescendantNodes(Location location, InvocationExpressionSyntax invocation)
+        {
+            // To optimise, we search first for the class constructor, then for the method declaration.
+            // If these cannot be found, we get the root of the syntax tree and search from there.
+            //
+            // This is needed for cases similar with:
+            // internal class VariousUsages
+            // {
+            //     XmlReader field = XmlReader.Create("uri", new XmlReaderSettings { DtdProcessing = DtdProcessing.Parse, XmlResolver = new XmlUrlResolver() });
+            // }
+            var root = location.SourceTree?.GetRoot()?.FindNode(location.SourceSpan) ??
+                       invocation.FirstAncestorOrSelf<MethodDeclarationSyntax>() ??
+                       invocation.SyntaxTree.GetRoot();
+
+            return root.DescendantNodes();
+        }
 
         private static bool IsXmlResolverDtdProcessingUnsafe(AssignmentExpressionSyntax assignment, SemanticModel semanticModel) =>
             semanticModel.GetConstantValue(assignment.Right).Value switch
             {
-                false => true,
+                false => true, // If ProhibitDtd is set to false the settings will be secure (there will be no parsing)
```

Comment:
```suggestion
                false => true, // If ProhibitDtd is set to false the settings will be unsafe (parsing is allowed)
```

Reply:
It seems I've messed up all the comments 😄